capsule/src/SessionManager.ts

```diff
@@ -1,29 +1,100 @@
 import userManagementClient from './UserManagementClient';
-import { SessionStorage } from './SessionStorage';
+import {SessionStorage} from './SessionStorage';
+import {
+  PublicKeyStatus,
+  PublicKeyType,
+} from '@usecapsule/user-management-client';
+import {InAppBrowser} from 'react-native-inappbrowser-reborn';
+import {
+  getWebAuthURLForCreate,
+  getWebAuthURLForLogin,
+  isSessionActive,
+} from './helpers';
 
 export default class SessionManager {
   private userId: string;
   private sessionStorage: SessionStorage;
+  private webauth: boolean;
   public async setSessionKey() {
-    return await userManagementClient.addSessionPublicKey(this.userId, {
-      publicKey: await this.sessionStorage.getPublicKey(),
-    });
+    const res = await userManagementClient.addSessionPublicKey(
+      this.userId,
+      this.webauth
+        ? {
+            status: PublicKeyStatus.PENDING,
+            type: PublicKeyType.WEB,
+          }
+        : {
+            publicKey: await this.sessionStorage.getPublicKey(),
+          }
+    );
+
+    if (!this.webauth) {
+      return res;
+    }
+    const link = getWebAuthURLForCreate(
+      res.data.id,
+      this.userId,
+      `${this.userId}@usecapsule.com`,
+      res.data.partnerId
+    );
+    console.log(link);
+
+    InAppBrowser.open(link);
+    while (true) {
+      if (await isSessionActive()) {
+        break;
+      }
+      console.log('Not active');
+      await new Promise((res) => setTimeout(res, 1000));
+    }
+    InAppBrowser.close();
+    console.log('active');
   }
 
-  constructor(userId: string, sessionStorage: SessionStorage) {
+  constructor(
+    userId: string,
+    sessionStorage: SessionStorage,
+    webauth: boolean
+  ) {
     this.userId = userId;
     this.sessionStorage = sessionStorage;
+    this.webauth = webauth;
   }
 
   public async refreshSessionIfNeeded() {
-    const challenge = await userManagementClient.getSessionChallenge(
-      this.userId
+    if (!this.webauth) {
+      const challenge = await userManagementClient.getSessionChallenge(
+        this.userId
+      );
+      const message = challenge.data.challenge;
+      const signature = await this.sessionStorage.signChallenge(message);
+      await userManagementClient.verifySessionChallenge(this.userId, {
+        signature,
+        publicKey: await this.sessionStorage.getPublicKey(),
+      });
+      return;
+    }
+
+    if (!(await isSessionActive())) {
+      return;
+    }
```

Comment:
why do we early return if session is not active?

Reply:
because then we don't need to extend it. This helper checks if it's active and if not, make some actions

Reply:
it looks like it only makes some actions if it is active?